firewall/inspection/portscan/detect.go

```diff
@@ -0,0 +1,357 @@
+		delete(ips, ipString)
+		return false
+	}
+
+	d.lastUpdated = time.Now()
+	return true
+}
+
+// Destroy implements the destroy interface.
+func (d *Detector) Destroy() error {
+	return nil
+}
+
+// DetectorFactory is a primitive detection method that runs within the factory only.
+func DetectorFactory(conn *network.Connection, pkt packet.Packet) (network.Inspector, error) {
+	return &Detector{}, nil
+}
+
+// Register registers the encryption detection inspector with the inspection framework.
+func init() {
+	ips = make(map[string]*ipData)
+
+	//cleanup old Threads
+	threads, _ := status.GetThreats(threadPrefix)
+	for _, t := range threads {
+		status.DeleteThreat(t.ID)
+	}
+
+	module = modules.Register("portscan-detection", nil, start, nil, "base", "netenv")
+	module.Enable()
+}
+
+func updateWholeList() {
+	log.Debugf("Portscan detection: update list&cleanup")
+	for ip, entry := range ips {
+		//done inside the loop to give other goroutines time in between to access the list (and during that time block this task)
+		runOnlyOne.Lock()
+		defer runOnlyOne.Unlock()
+
+		if entry.updateScoreIgnoreBlockPrevOffender(ip) {
+			log.Debugf("%s: %v", ip, entry)
+		} else {
+			log.Debugf("Removed %s from the list", ip)
+		}
+	}
+	log.Debugf("Portscan detection: finished update list&cleanup")
+
+}
+
+func start() (err error) {
+	go delayedStart()
+
+	// Reload own IP List on Network change
+	err = module.RegisterEventHook(
+		"netenv",
+		"network changed",
+		"Reload List of own IPs on Network change for Portscan detection",
+		func(_ context.Context, _ interface{}) (err error) {
+			fillOwnIPs()
+
+			return
+		},
+	)
+
+	fillOwnIPs()
+
+	return
+}
+
+func delayedStart() {
+	time.Sleep(startAfter)
+
+	log.Debugf("starting Portscan detection")
+	err := inspection.RegisterInspector(&inspection.Registration{
+		Name:    "Portscan Detection",
+		Order:   0,
+		Factory: DetectorFactory,
+	})
+
+	if err != nil {
+		panic(err)
+	}
+
+	module.NewTask("portscan score update", func(ctx context.Context, task *modules.Task) (err error) {
+		updateWholeList()
+		return
+	}).Repeat(cleanUpInterval).MaxDelay(cleanUpMaxDelay)
+}
+
+func fillOwnIPs() {
+	var err error
+	ownIPs, _, err = netenv.GetAssignedAddresses()
+
+	if err != nil {
+		log.Errorf("Couldn't obtain List of IPs: %v", err)
+	}
+
+	log.Debugf("Portscan detection: ownIPs: %v", ownIPs)
+}
+
+//Does NOT check localhost range!!
+func foreignIPv4(ip net.IP) bool {
+	if ip.To4() == nil {
+		return false
+	}
+
+	for _, ownIP := range ownIPs {
+		if ip.Equal(ownIP) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isNetBIOSoverTCPIP(conn *network.Connection) bool {
+	return conn.LocalPort == 138 ||
+		(conn.IPProtocol == packet.UDP && conn.LocalPort == 138) ||
+		(conn.IPProtocol == packet.TCP && conn.LocalPort == 139)
+
+}
+
+// Source: https://stackoverflow.com/questions/27516387/what-is-the-correct-way-to-find-the-min-between-two-integers-in-go#27516559
```

Comment:
Nice that you quote the references but they blow up the code and IMHO we can remove this one